lib/nolibc/stdio.c

```diff
@@ -541,3 +544,153 @@ int puts(const char *s)
 {
 	return fputs_internal(s, stdout, 1);
 }
+
+#if CONFIG_LIBVFSCORE
+struct _nolibc_file {
+	int fd;
+	bool eof;
+};
+
+/* The following code is derived from musl libc */
+static int __fmodeflags(const char *mode, int *flags)
+{
+	if (!strchr("rwa", *mode)) {
+		errno = EINVAL;
+		return -1;
+	}
+	if (strchr(mode, '+'))
+		*flags = O_RDWR;
+	else if (*mode == 'r')
+		*flags = O_RDONLY;
+	else
+		*flags = O_WRONLY;
+	if (strchr(mode, 'x'))
+		*flags |= O_EXCL;
+	if (strchr(mode, 'e'))
+		*flags |= O_CLOEXEC;
+	if (*mode != 'r')
+		*flags |= O_CREAT;
+	if (*mode == 'w')
+		*flags |= O_TRUNC;
+	if (*mode == 'a')
+		*flags |= O_APPEND;
+
+	return 0;
+}
+
+FILE *fopen(const char *pathname, const char *mode)
+{
+	int flags;
+	int fd;
+	int ret;
+
+	ret = __fmodeflags(mode, &flags);
+
+	if (ret < 0)
+		return NULL;
+
+	fd = open(pathname, flags, 0666);
+
+	if (fd < 0)
+		return NULL;
+
+	return fdopen(fd, mode);
+}
+
+int feof(FILE *stream)
+{
+	return stream->eof;
+}
+
+int fclose(FILE *stream)
+{
+	int ret = close(stream->fd);
+
+	free(stream);
+	return ret;
+}
+
+FILE *fdopen(int fd, const char *mode __unused)
+{
+	FILE *f = (FILE *)malloc(sizeof(FILE));
+
+	if (!f)
+		return NULL;
+	f->fd = fd;
+	f->eof = 0;
+	return f;
+}
+
+int fseek(FILE *stream, long offset, int whence)
+{
+	return lseek(stream->fd, offset, whence);
+}
+
+size_t fread(void *ptr, size_t size, size_t nmemb, FILE *stream)
+{
+	if (unlikely(!stream || !ptr || !size || !nmemb)) {
```

Comment:
Can you also `#include <uk/arch/lcpu.h>` to make sure the `unlikely` macro is defined? At the moment it should be indirectly be available but it is always good to explicitly include things in case nothing breaks if indirect headers are changed.